core/src/main/java/org/ehcache/Ehcache.java

```diff
@@ -263,6 +267,7 @@ public V apply(final K key, final V previousValue) {
         }
       } finally {
         putObserver.end(PutOutcome.FAILURE);
+        eventNotificationService.fireSuccessfulEvents();
```

Comment:
This `finally` is inside the `catch` block. So there are ways to exit this method without cleaning / firing events. This feels very brittle to me. We should _ALWAYS_ handle events one way or another.